### ISSUE TITLE
[a11y deliverable] Audit: Chatbot MHV Prescriptions Skill - Findings

### ISSUE THREAD

**Issue by sara-amanda** [department-of-veterans-affairs/va.gov-team#67376]
# Audit Findings: Chatbot MHV Prescriptions Skill 

**External consideration:** We realize that some of the issues identified with the chatbot may be specific to Microsoft, beyond the scope and ability of the VA to remediate, and that tickets may need to be drawn up with Microsoft. This audit was completed to bring attention to all issues, regardless of owner, for a comprehensive accessibility review of the chatbot for the end user. 

Should you have any questions regarding this deliverable, please let us know. The CAIA A11ys are happy to help! 

## In this deliverable
**Below you will find the following sections:** 
- Initial Chatbot Findings
- General Chatbot Findings
- Specific Chatbot Findings
- Audit Spreadsheet
- Audit Files Attached/Linked
- Audit Snapshot

## Initial Chatbot Findings
These findings have previously been shared; however, we are aggregating them here for quick reference.
- [CAIA Initial Video Demo Feedback for Chatbot](https://github.com/department-of-veterans-affairs/va.gov-team/issues/65945#issuecomment-1731568653) (9/22/2023)
- [CAIA A11y Review of Chatbot on dev.va.gov](https://github.com/department-of-veterans-affairs/va.gov-team/issues/65945#issuecomment-1736079347) (9/26/2023)
- [CAIA Update: Launch-Blocking Issue](https://github.com/department-of-veterans-affairs/va.gov-team/issues/65945#issuecomment-1739582767) (9/28/2023)

## General Chatbot Findings
The following are links to CAIA a11y ticket comments regarding general findings with the Chatbot that describe the issue, illustrates the issue with images and/or screen recordings, and provides background on the testing details. 

- [Keyboard Trap](https://github.com/department-of-veterans-affairs/va.gov-team/issues/66754#issuecomment-1753263632)
- [Ambiguous UX](https://github.com/department-of-veterans-affairs/va.gov-team/issues/66754#issuecomment-1757751016)
- [Multi-Modal Input Methods](https://github.com/department-of-veterans-affairs/va.gov-team/issues/66754#issuecomment-1757879905)
- [Message Sequence/AT](https://github.com/department-of-veterans-affairs/va.gov-team/issues/66754#issuecomment-1757894655)

## Specific Chatbot Findings
The following are specific Chatbot findings, as a result of the audit testing plan put in place. Additional information will be shared via the audit spreadsheet, to supplement these findings. 
- #67209
- #67286
- #67315
- #67386
- #67309 

## Audit Spreadsheet

This spreadsheet combines information from multiple sources such as the Microsoft VPAT, WCAG and COGA, to provide a comprehensive audit of the chatbot. CAIA feedback can be found in the yellow columns, listed below. 

- Tester
- Impact Area(s) RX/Chatbot
- Pass/Fail
- Severity/Priority from Audit
- Issue Location(s)
- GitHub Ticket Link (links can also be found at the top of this ticket)
- Steps to Reproduce (also provided in the GitHub tickets)
- Notes

### Audit Files Attached/Linked

**Please refer to the `Failed Tab` to view items that did not pass.** 

- [**Google Sheet:** 2023 10 11 Deliverable Chatbot Audit - CAIA A11ys](https://docs.google.com/spreadsheets/d/1KY_apLWtdDMxbA42sfuDqC81umA3yF1Yt5NwXZqbEYE/edit?usp=sharing)
- [**Excel:** 2023 10 11 Deliverable Chatbot Audit - CAIA A11ys.xlsx](https://github.com/department-of-veterans-affairs/va.gov-team/files/12874603/2023.10.11.Deliverable.Chatbot.Audit.-.CAIA.A11ys.xlsx)

### Audited Items

The **CAIA Accessibility Team** successfully audited 65.45% of auditable items on the CAIA a11y Chatbot Audit Spreadsheet using the Dev environment provided for the Chatbot MHV Prescriptions Skill Feature.

Based on the date of staging (10/12/2023) and the findings found thus far, we are going to pause our audit and share these findings with the Governance Team and the Project Team. 

**Audit Status** | **Audit Items (Count)** | **Percent of Total (%)**
-- | -- | --
Failed | 9 | 16.36% 
Further investigation needed | 5 | 9.09%
Passed | 19 | 34.55%
Addressed in another recommendation | 3 | 5.45%
Remaining items be audited | 19 | 34.55%
**Total** | **55** | **100.00%**

### Finding Types
Since the chatbot is a different flow than webpages or a form, we are testing by using WCAG and W3C’s Cognitive and Learning Disabilities Accessibility Task Force (COGA TF) to fill in the gaps directed at cognitive disabilities. 

The items that failed were equally in relation to WCAG and COGA: 

**Type** | **Count of Audited Items**
-- | -- 
COGA Only | 3
Not WCAG or COGA | 2
WCAG and COGA | 1
WCAG only | 3
**Total Findings Failing** | 9

### Impact of Impact Areas

**Impact Areas** | **Impact of Appearance (Count)** | **Percent of Impact Areas (%)**
-- | -- | --
Cognitive | 14 | 48.28%
Visual | 8 | 27.59%
Auditory | 2 | 6.90%
Speech | 2 | 6.90%
Physical | 3 | 10.34%
**Total** | **29** | **100.00%**

_The above totals reflect duplicate tickets if the issue audited appeared in one or more category._

**Comment by sara-amanda** [comment]
## Shared Deliverable
[This deliverable has been shared](https://dsva.slack.com/archives/C01KTS3F493/p1697059815335249) with both the product team and Governance on 10/11/2023 via slack in the DSVA #va-virtual-agent-public channel. Thank you, @coforma-terry!

**Comment by Joy-Elizabeth** [comment]
Chatbot team does have these final two audit findings as a priority to address post Rx Skill launch. We are dependent on MSFT and will work closely with them to complete:

https://github.com/department-of-veterans-affairs/va.gov-team/issues/67386
https://github.com/department-of-veterans-affairs/va.gov-team/issues/67309

**Comment by sara-amanda** [comment]
## CAIA Update

- Closing this ticket, since the deliverable has been shared.
    - However, the deliverable in this ticket is still available for access in the closed state.
- A copy also resides [in CAIA's GitHub Deliverables Folder for VA Chatbot](https://github.com/department-of-veterans-affairs/va.gov-team/edit/master/teams/CAIA/accessibility/Deliverables/VA%20Chatbot/2023%2010%2011%20Findings%20VA%20Chatbot%20Audit%3A%20MHV%20Prescriptions%20Skill.md).

**Comment by anitadewitt-va** [comment]
**Regarding the keyboard trap:**
We have not been able to recreate this issue. Upon tabbing into the chat window, the user is presented with the keyboard help modal. One additional tab closes the modal and focuses the chat history. The help modal can then be re-triggered with Tab+Shift and closed with another tab whenever the user would like to see the keyboard modal.

**Keyboard navigation findings:**

- Technically the users "can navigate to all chatbot content and interactive elements using the keyboard". 
- In order to access response buttons, a user can tab until the chat history is selected, then use the arrow keys to focus on newer/older messages.
- When a message with interactive elements is in focus the user is able to press Enter to shift focus into the various elements contained on the card. This creates a "focus trap" where the user can tab through each element in the card until pressing ESC and focus is returned to the chat history.
- Users can also use Shift+Tab to tab up/backwards from the text input to select option buttons when they appear.

**From [BotFramework-WebChat/docs/accessibility.md](https://github.com/microsoft/BotFramework-WebChat/blob/main/docs/ACCESSIBILITY.md#ux-keyboard-help-screen)**
[Private Zenhub Image](https://api.zenhub.com/attachedFiles/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBL0UrQVE9PSIsImV4cCI6bnVsbCwicHVyIjoiYmxvYl9pZCJ9fQ==--425f9e843a358625dad91a89a1770f98695ac5bd/image.png)

**Comment by sara-amanda** [comment]
## CAIA Office Hours

### 10/31/2023
- **Attendees:**  `1` Team: Chatbot Team
- **Support types:** `Accessibility`

<details><summary>Toggle to view an overview of a11y-related support given on 10/31/2023</summary>

- **CAIA Presence**: `Sara`and `Lily`
- **Group Needing Support's Slack Channel:** [va-virtual-agent-public ](https://dsva.slack.com/archives/C01KTS3F493)
- **Resources Exchanged:** 
    -  #67376 
    - [Links to the Chatbot Playbook in Column A of the Deliverable](https://docs.google.com/spreadsheets/d/1KY_apLWtdDMxbA42sfuDqC81umA3yF1Yt5NwXZqbEYE/edit#gid=98119274)
    - **Chatbot Team** [Chatbot Team Ticket #1201](https://github.com/department-of-veterans-affairs/va-virtual-agent/issues/1201). 
        - **Asking CAIA:** Would the [Microsoft recommendation](https://github.com/microsoft/BotFramework-WebChat/blob/main/docs/ACCESSIBILITY.md#ux-navigating-activities-in-the-chat-history) of the [Grid Layout Widget](https://www.w3.org/WAI/ARIA/apg/patterns/grid/) be applicable here (see W3 standards).
    - #67309 
    - **Screen Recordings:** Anita added a comment with the [recordings](https://github.com/department-of-veterans-affairs/va.gov-team/issues/67309#issuecomment-1788042975) they showed on 10/31/2023 [#67309](https://github.com/department-of-veterans-affairs/va.gov-team/issues/67309).

#### Topics Discussed: 
- Progress on the Accessibility Defects
- Challenges Chatbot Team is Facing
- Clarifications and questions answered

</details>

<details><summary>Toggle to view an expanded view of a11y-related notes from 10/31/2023</summary>

#### Keyboard Navigation

- **Scroll with a mouse and scroll up and down smoothly with arrow keys.** 
    - Similar functionality in our chatbot. 
    - Keyboard navigation instead of the smooth scroll it jumps from message to message. 
    - **Not their expectations** - _Governance Team_ 
- **Use a semantic tool to look at Dev tools to mitigate those issues.** 
    - **Raina**: Recommendations that were given from the findings, made those changes to see if we could achieve the desired behavior. Using some of the pattern for a grid widget, that is why we are able to use up and down arrow keys in the chat history. 
    
> **Question**
 _Since Microsoft has done research on best practices, on applications similar to our chatbot. Is that a guideline that we could be following?_
>- Chatbot Team [Chatbot Team Ticket #1201](https://github.com/department-of-veterans-affairs/va-virtual-agent/issues/1201) 
>- Would the [Microsoft recommendation](https://github.com/microsoft/BotFramework-WebChat/blob/main/docs/ACCESSIBILITY.md#ux-navigating-activities-in-the-chat-history) of the [Grid Layout Widget](https://www.w3.org/WAI/ARIA/apg/patterns/grid/) be applicable here? (see W3 standards). 

#### Reading/Skipping Messages
##### VoiceOver // Windows NVDA // JAWS 
- [CAIA Card Ticket #67309 related issue](https://github.com/department-of-veterans-affairs/va.gov-team/issues/67309): Instead of reading every message out, the screen reader for VO will not read them one at a time, as they are appearing. 
    - Microsoft would be submitting a ticket to VoiceOver. 
    - They can look into it. 
    - Microsoft is not able to replicate it.
    - Inconsistencies within VoiceOver using the same flow. 

#### Adaptive Cards // Larger Blocks
- **How they are being read out … have implemented some changes:** 
    - Decided to add a delay based on the number of characters in the message itself. 
    - Depending on how long the message is, added a slight delay, increase the delay in between messages, when a bubble appears, to give the user more time to understand the message delivered by the SR. Give them more time to react to the message as well, if certain actions need to be delivered. 
    - This helps to get around messages appearing too fast. 
    - The next change, for the adaptive card, they have added a summary to that card, so that it actually displays key characteristics, summary to let the SR know that a card is being displayed on the screen, use keyboard to get elements within the card. 
    - It improves the experience. 
    - **Video from Anita shown.** 
        - Display message → Delay → adaptive card → delayed 
        - Separate outputs with delay in between
        - It takes longer for the message to be there. 
        - Summary of the adaptive card reading is off a bit, previously it would have been skipped. 
            - Speech synthesis markup language being read. 
                - That same tag is being used by the SR. 
                - Jaws reads that correctly, VO does not, and reads it as intended.) 
                - There is some inconsistency across screen readers, the speak tag of the adaptive card, and how it works across SR.
- **Microsoft’s Expectations vs VA Expectations how to navigate:** 
    - Summary of the card, but you would need for keyboard navigation to dive into the card to be able to get to each of the fields. 
    - Speak tag to the card as a whole, card and short description, like a caption on an image as a summary. The intent of the speak tag is not to expose every text content in the card itself. Expectation is that if the user wants more details and what labels and values that they can use the keyboard to navigate in the chat history to get into the card to get the detail. 
    - _**Functionality as expressed by Microsoft. ^**_
    - **Additional information** - originally this functionality was for a voice experience with no screen at all. Screen you got a card if no card summary text above and ability to ask follow-up questions. This functionality still exists. The experience can stand without that card being read out on a screen reader because the reader has the ability to ask other questions. 
    - No Microsoft Documentation - Mark Tucker. 
    - **_Read off one at a time so that navigation only goes to the previous one, in the event there are multiple prescriptions._** 

#### Questions
- **Is it possible to create a VoiceOver Transcript?**
- **Overarching Categories (team is thinking this through further - how to approach):**
    - Face-Value Feedback
    - Limitations of Microsoft 
    - VA Standards
- **Next Time: Chatbot Team Will** ... 
    - Go over the cards and come back with the essence that is remaining … Microsoft or CAIA support. 
    - Level 1 Feedback Impact so far? 
    - Defect at face value - Microsoft’s feedback on behavior and how it relates to the VA standard. 
- **Current Approach by Chatbot Team**
    - Overarching themes SR and Keyboard Navigation Issues
    - Are there any other overarching umbrellas for feedback to see if anything is missing.
- **Should we consider the W3 Patterns, based on feedback from Microsoft?** 
- **Who owns the VA Accessibility Standards?** (DST/USWDS) 
</details>

## CAIA Sync on Chatbot
- CAIA is going to meet and discuss the items mentioned in Office Hours on 11/1/2023.
- CAIA will update the Chatbot Team, with feedback when finalized.
